### PR TITLE
Flash simulator enhancements

### PR DESCRIPTION
**Pull request by aescolar**
For the POSIX architecture, add options to:
* Clear the flash content at boot
* To delete the file on exit
* To just keep the flash in RAM instead of a file in disk

It is normal for users to want to do this things, let's just provide them command line options, so they don't need to delete the files by hand before/after running their test.

### CODE REVIEW

File: drivers/flash/flash_simulator.c [hunk 1]
```diff
@@ -147,7 +149,11 @@ STATS_NAME_END(flash_sim_thresholds);
 static uint8_t *mock_flash;
 static int flash_fd = -1;
 static const char *flash_file_path;
-static const char default_flash_file_path[] = "flash.bin";
+#define DEFAULT_FLASH_FILE_PATH "flash.bin"
```
Comment (de-nordic) [inline drivers/flash/flash_simulator.c:152]: I think this should be moved somewhere on top of the file, under includes, as a visibly defined default.
Having it here to just assign it to something in  the next line and not using it anywhere seems to be just waste of line.
This should be more clearly advertised as something that impacts internals of the source.

File: drivers/flash/flash_simulator.c [hunk 2]
```diff
@@ -372,46 +378,58 @@ static int flash_mock_init(const struct device *dev)
...
-	if (flash_fd == -1) {
-		posix_print_warning("Failed to open flash device file "
-				    "%s: %s\n",
-				    flash_file_path, strerror(errno));
-		return -EIO;
-	}
+	if (flash_on_ram == true) {
+		mock_flash = (uint8_t *)malloc(FLASH_SIMULATOR_FLASH_SIZE);
+		if (mock_flash == NULL) {
+			posix_print_warning("Could not allocate flash in Linux process heap %s\n",
```
Comment (de-nordic) [inline drivers/flash/flash_simulator.c:386]: This line incorrectly names my FreeBSD.

File: drivers/flash/flash_simulator.c [hunk 3]
```diff
@@ -147,7 +149,11 @@ STATS_NAME_END(flash_sim_thresholds);
 static uint8_t *mock_flash;
 static int flash_fd = -1;
 static const char *flash_file_path;
-static const char default_flash_file_path[] = "flash.bin";
+#define DEFAULT_FLASH_FILE_PATH "flash.bin"
+static const char default_flash_file_path[] = DEFAULT_FLASH_FILE_PATH;
+static bool flash_erase_at_start;
+static bool flash_rm_at_exit;
+static bool flash_on_ram;
```
Comment (de-nordic) [inline drivers/flash/flash_simulator.c:156]: In RAM?

File: drivers/flash/flash_simulator.c [hunk 4]
```diff
@@ -372,46 +378,58 @@ static int flash_mock_init(const struct device *dev)
...
 
-	rc = fstat(flash_fd, &f_stat);
-	if (rc) {
-		posix_print_warning("Failed to get status of flash device file "
-				    "%s: %s\n",
-				    flash_file_path, strerror(errno));
-		return -EIO;
-	}
+		if (flash_file_path == NULL) {
+			flash_file_path = default_flash_file_path;
```
Comment (de-nordic) [inline drivers/flash/flash_simulator.c:393]: Seems that we can just assign here `DEFAULT_FLASH_FILE_PATH`  and drop the `default_flash_file_path` completely.

File: drivers/flash/flash_simulator.c [hunk 5]
```diff
@@ -372,46 +378,58 @@ static int flash_mock_init(const struct device *dev)
...
+			posix_print_warning("Failed to mmap flash device file "
+					"%s: %s\n",
+					flash_file_path, strerror(errno));
+			return -EIO;
+		}
 	}
 
-	if (f_stat.st_size == 0) {
-		/* erase the memory unit by pulling all bits to one */
+	if ((flash_erase_at_start == true) || (flash_on_ram == true) || (f_stat.st_size == 0)) {
```
Comment (de-nordic) [inline drivers/flash/flash_simulator.c:429]: So erase happens by default for RAM?

**Review by de-nordic — COMMENTED** [review]
Generally looks OK.
Some minor comments and missing Nordic copyright updates.

File: drivers/flash/flash_simulator.c (hunk 2 above)
Comment (aescolar) [reply drivers/flash/flash_simulator.c:386]: :)

File: drivers/flash/flash_simulator.c (hunk 4 above)
Comment (aescolar) [reply drivers/flash/flash_simulator.c:393]: flash_file_path is just a pointer. So we need something to store that default string.

File: drivers/flash/flash_simulator.c (hunk 5 above)
Comment (aescolar) [reply drivers/flash/flash_simulator.c:429]: Yep. Just like when the flash simulator runs in other archs. I will add a comment in the command line option.

File: drivers/flash/flash_simulator.c (hunk 1 above)
Comment (aescolar) [reply drivers/flash/flash_simulator.c]: done (note it is used below in the command line argument description)

File: drivers/flash/flash_simulator.c (hunk 4 above)
Comment (de-nordic) [reply drivers/flash/flash_simulator.c:394]: > flash_file_path is just a pointer. So we need something to store that default string.

The `default_flash_file_path` is also just a pointer here.
`DEFAULT_FLASH_FILE_PATH` is name for literal string. You decide its fate by assigning it to something or not. Not assigned is just left out, assigned to `const char *` or `const static char[]` type is placed in ro section of binary, assigned to `static char []` type is placed in writable .data section. Space for it is reserved the moment you reference it.

Comment (aescolar) [reply drivers/flash/flash_simulator.c]: Thank you. Change done.

**Review by daor-oti — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Pull request merged**